src/actions/plugins/imswitcher.ts

```diff
@@ -0,0 +1,144 @@
+    ) {
+      if (
+        newMode !== ModeName.Insert &&
+        newMode !== ModeName.SurroundInputMode &&
+        newMode !== ModeName.Replace
+      ) {
+        this.switchToDefaultIM();
+        return;
+      }
+    }
+    // when you enter insert-like mode, resume origin input method
+    if (
+      newMode === ModeName.Insert ||
+      newMode === ModeName.SurroundInputMode ||
+      newMode === ModeName.Replace
+    ) {
+      if (
+        prevMode !== ModeName.Insert &&
+        prevMode !== ModeName.SurroundInputMode &&
+        prevMode !== ModeName.Replace
+      ) {
+        this.resumeIM();
+      }
+    }
+  }
+
+  // save origin input method and set input method to default
+  private async switchToDefaultIM() {
+    const obtainIMCmd = configuration.autoSwitchInputMethod.obtainIMCmd;
+    const rawObtainIMCmd = this.getRawCmd(obtainIMCmd);
+    if (existsSync(rawObtainIMCmd) || Globals.isTesting) {
+      try {
+        const insertIMKey = await this.execute(obtainIMCmd);
+        if (insertIMKey !== undefined) {
+          this.savedIMKey = insertIMKey.trim();
+        }
+      } catch (e) {
+        logger.error(`IMSwitcher: promise is rejected. err=${e}`);
+      }
+    } else {
+      this.showCmdNotFoundErrorMessage(rawObtainIMCmd, 'vim.autoSwitchInputMethod.obtainIMCmd');
+    }
+
+    const defaultIMKey = configuration.autoSwitchInputMethod.defaultIM;
+    if (defaultIMKey !== this.savedIMKey) {
+      this.switchToIM(defaultIMKey);
+    }
+  }
+
+  // resume origin inputmethod
+  private async resumeIM() {
+    if (this.savedIMKey !== configuration.autoSwitchInputMethod.defaultIM) {
+      this.switchToIM(this.savedIMKey);
+    }
+  }
+
+  private async switchToIM(imKey: string) {
+    let switchIMCmd = configuration.autoSwitchInputMethod.switchIMCmd;
+    const rawSwitchIMCmd = this.getRawCmd(switchIMCmd);
+    if (existsSync(rawSwitchIMCmd) || Globals.isTesting) {
+      if (imKey !== '' && imKey !== undefined) {
+        switchIMCmd = switchIMCmd.replace('{im}', imKey);
+        try {
+          await this.execute(switchIMCmd);
+        } catch (e) {
+          logger.error(`IMSwitcher: promise is rejected. err=${e}`);
+        }
+      }
+    } else {
+      this.showCmdNotFoundErrorMessage(rawSwitchIMCmd, 'vim.autoSwitchInputMethod.switchIMCmd');
+    }
+  }
+
+  private getRawCmd(cmd: string): string {
+    return cmd.split(' ')[0];
+  }
+
+  private showCmdNotFoundErrorMessage(cmd: string, config: string) {
+    Message.ShowError('Unable to find ' + cmd + '. check your ' + config + ' in VSCode setting.');
+    this.disableIMSwitch();
+  }
+
+  private disableIMSwitch() {
+    const originConfig = configuration.autoSwitchInputMethod;
+    configuration.autoSwitchInputMethod = {
+      enable: false,
+      defaultIM: originConfig.defaultIM,
+      switchIMCmd: originConfig.switchIMCmd,
+      obtainIMCmd: originConfig.obtainIMCmd,
+    };
+  }
+
+  private validateConfiguration(): boolean {
```

Comment:
Nitpick: as we are returning a boolean, can you rename this method to `isValidateConfiguration()`